kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java

```diff
@@ -290,8 +412,121 @@ private static boolean hasPhysicalName(StructField field) {
     }
 
     private static int getColumnId(StructField field) {
-        // TODO: add a method to FieldMetadata to extract this as an int
-        // instead of having to parse it here
-        return Integer.parseInt(field.getMetadata().get(COLUMN_MAPPING_ID_KEY).toString());
+        return field.getMetadata().getLong(COLUMN_MAPPING_ID_KEY).intValue();
+    }
+
+    private static boolean hasNestedColumnIds(StructField field) {
+        return field.getMetadata().contains(COLUMN_MAPPING_NESTED_IDS_KEY);
+    }
+
+    private static FieldMetadata getNestedColumnIds(StructField field) {
+        return field.getMetadata().getMetadata(COLUMN_MAPPING_NESTED_IDS_KEY);
+    }
+
+    private static int getMaxNestedColumnId(StructField field) {
+        return getNestedColumnIds(field).getEntries().values().stream()
+                .filter(v -> v instanceof Long)
+                .map(v -> (Long) v)
+                .max(Long::compare)
+                .orElse(0L)
+                .intValue();
+    }
+
+    private static StructField transformSchema(
+            AtomicInteger currentFieldId,
+            StructField structField,
+            String path) {
+        DataType dataType = structField.getDataType();
+        if (dataType instanceof StructType) {
+            StructType type = (StructType) dataType;
+            List<StructField> fields =
+                    type.fields().stream()
+                            .map(field -> {
+                                StructField nestedField = transformSchema(
+                                        currentFieldId,
+                                        field,
+                                        getPhysicalName(field));
+                                return new StructField(
+                                        nestedField.getName(),
+                                        nestedField.getDataType(),
+                                        nestedField.isNullable(),
+                                        nestedField.getMetadata());
+                            })
+                            .collect(Collectors.toList());
+            return new StructField(
+                    structField.getName(),
+                    new StructType(fields),
+                    structField.isNullable(),
+                    structField.getMetadata());
+        } else if (dataType instanceof ArrayType) {
+            ArrayType type = (ArrayType) dataType;
+            path = "".equals(path) ? getPhysicalName(structField) : path;
```

Comment:
Can you avoid updating the input argument? It made me think for a second. If you use `String baseElementPath` and `finalElementPath` that could be clearer

Reply:
done